source/text/Font.cpp

```diff
@@ -14,573 +14,823 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+	pixelSize = size;
+	UpdateFont();
+}
+
+
+
+void Font::SetDrawingSettings(const DrawingSettings &newSettings)
+{
+	drawingSettings = newSettings;
+	UpdateFont();
 }
 
 
 
 void Font::Draw(const DisplayText &text, const Point &point, const Color &color) const
 {
-	DrawAliased(text, round(point.X()), round(point.Y()), color);
+	DrawCommon(text, point.X(), point.Y(), color, true);
 }
 
 
 
 void Font::DrawAliased(const DisplayText &text, double x, double y, const Color &color) const
 {
-	int width = -1;
-	const string truncText = TruncateText(text, width);
-	const auto &layout = text.GetLayout();
-	if(width >= 0)
-	{
-		if(layout.align == Alignment::CENTER)
-			x += (layout.width - width) / 2;
-		else if(layout.align == Alignment::RIGHT)
-			x += layout.width - width;
-	}
-	DrawAliased(truncText, x, y, color);
+	DrawCommon(text, x, y, color, false);
 }
 
 
 
 void Font::Draw(const string &str, const Point &point, const Color &color) const
 {
-	DrawAliased(str, round(point.X()), round(point.Y()), color);
+	DrawCommon({str, {}}, point.X(), point.Y(), color, true);
 }
 
 
 
 void Font::DrawAliased(const string &str, double x, double y, const Color &color) const
 {
-	glUseProgram(shader.Object());
-	glBindTexture(GL_TEXTURE_2D, texture);
-	glBindVertexArray(vao);
-	
-	glUniform4fv(colorI, 1, color.Get());
-	
-	// Update the scale, only if the screen size has changed.
-	if(Screen::Width() != screenWidth || Screen::Height() != screenHeight)
-	{
-		screenWidth = Screen::Width();
-		screenHeight = Screen::Height();
-		GLfloat scale[2] = {2.f / screenWidth, -2.f / screenHeight};
-		glUniform2fv(scaleI, 1, scale);
-	}
-	
-	GLfloat textPos[2] = {
-		static_cast<float>(x - 1.),
-		static_cast<float>(y)};
-	int previous = 0;
-	bool isAfterSpace = true;
-	bool underlineChar = false;
-	const int underscoreGlyph = max(0, min(GLYPHS - 1, '_' - 32));
-	
-	for(char c : str)
-	{
-		if(c == '_')
-		{
-			underlineChar = showUnderlines;
-			continue;
-		}
-		
-		int glyph = Glyph(c, isAfterSpace);
-		if(c != '"' && c != '\'')
-			isAfterSpace = !glyph;
-		if(!glyph)
-		{
-			textPos[0] += space;
-			continue;
-		}
-		
-		glUniform1i(glyphI, glyph);
-		glUniform1f(aspectI, 1.f);
-		
-		textPos[0] += advance[previous * GLYPHS + glyph] + KERN;
-		glUniform2fv(positionI, 1, textPos);
-		
-		glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
-		
-		if(underlineChar)
-		{
-			glUniform1i(glyphI, underscoreGlyph);
-			glUniform1f(aspectI, static_cast<float>(advance[glyph * GLYPHS] + KERN)
-				/ (advance[underscoreGlyph * GLYPHS] + KERN));
-			
-			glUniform2fv(positionI, 1, textPos);
-			
-			glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
-			underlineChar = false;
-		}
-		
-		previous = glyph;
-	}
-	
-	glBindVertexArray(0);
-	glUseProgram(0);
+	DrawCommon({str, {}}, x, y, color, false);
 }
 
 
 
-int Font::Width(const string &str, char after) const
+int Font::Width(const string &str) const
 {
-	return WidthRawString(str.c_str(), after);
+	return TextFromViewCeilX(ViewWidth({str, {}}));
```

Comment:
Render() caches the result, including the texture and height. Next Draw() uses the cached texture because it specifies no layout parameter except one place in Table.cpp.